### PR TITLE
2020 04 08 multi db dao

### PR DESCRIPTION
**Pull request by Christewart**
Begins to implement changes required for #1240 

This PR does _not_ implement a new database, it is only meant to re-architect our existing infrastructure to support new databases in the future. The goal is to have all unit tests pass with this new architecture.

https://scala-slick.org/doc/3.3.1/database.html#databaseconfig

https://github.com/slick/slick/tree/master/samples/slick-multidb/src/main/scala/cake

The basic idea is moving the `Table.scala` file inside of the `DAO` trait so they can share the same `profile.api._`

### CODE REVIEW

**Comment by benthecarman** [comment]
Building the Docs is failing

**Comment by benthecarman** [comment]
Should rename all the `XTable.scala` files to `XDb.scala` since they now only contain the Db representations

File: db-commons/src/main/scala/org/bitcoins/db/DbManagement.scala [hunk 1]
```diff
@@ -1,33 +1,28 @@
...
   /** Creates all tables in our table list, in one SQL transaction */
-  def createAll()(
-      implicit config: AppConfig,
-      ec: ExecutionContext
-  ): Future[Unit] = {
-    val tables = allTables.map(_.baseTableRow.tableName).mkString(", ")
-    logger.debug(s"Creating tables: $tables")
+  def createAll()(implicit ec: ExecutionContext): Future[Unit] = {
+    /*    val tables = allTables.map(_.baseTableRow.tableName).mkString(", ")
+    logger.debug(s"Creating tables: $tables")*/
```
Comment (benthecarman) [inline db-commons/src/main/scala/org/bitcoins/db/DbManagement.scala]: ping

File: node/src/main/scala/org/bitcoins/node/db/NodeDbManagement.scala [hunk 2]
```diff
@@ -1,12 +1,23 @@
...
+trait NodeDbManagement extends DbManagement { _: JdbcProfileComponent =>
+
+  import profile.api._
+
+  def ec: ExecutionContext
+
+  override def appConfig: NodeAppConfig
+
+  private lazy val txTable: TableQuery[Table[_]] = {
+    BroadcastAbleTransactionDAO()(appConfig.asInstanceOf[NodeAppConfig], ec).table
```
Comment (benthecarman) [inline node/src/main/scala/org/bitcoins/node/db/NodeDbManagement.scala]: `appConfig` is already an instance of `NodeAppConfig`

File: node/src/main/scala/org/bitcoins/node/models/BroadcastAbleTransactionDAO.scala [hunk 3]
```diff
@@ -26,4 +30,29 @@ final case class BroadcastAbleTransactionDAO(profile: JdbcProfile)(
...
+      Some(tx.transaction.txId.flip, tx.transaction.bytes, tx.id)
+
+    import org.bitcoins.db.DbCommonsColumnMappers._
+
+    def txid: Rep[DoubleSha256DigestBE] = column("txid", O.Unique)
+    def bytes: Rep[ByteVector] = column("tx_bytes")
+
+    def * : ProvenShape[BroadcastAbleTransaction] =
+      (txid, bytes, id.?) <>
+        (fromTuple, toTuple)
```
Comment (benthecarman) [inline node/src/main/scala/org/bitcoins/node/models/BroadcastAbleTransactionDAO.scala:55]: In the `BlockHeaderDAO` you were able to get rid of the `fromTuple` and `toTuple` design, can you do that here and the other DAOs. Could then close #1137

File: node/src/main/scala/org/bitcoins/node/models/PeerDAO.scala [hunk 4]
```diff
@@ -1,12 +0,0 @@
-package org.bitcoins.node.models
-
-import org.bitcoins.db.{CRUDAutoInc}
-import slick.jdbc.SQLiteProfile.api._
-
-import scala.concurrent.ExecutionContext
-import org.bitcoins.node.config.NodeAppConfig
-
-case class PeerDAO()(implicit ec: ExecutionContext, appConfig: NodeAppConfig)
```
Comment (benthecarman) [inline node/src/main/scala/org/bitcoins/node/models/PeerDAO.scala:9]: Are we removing this table altogether?

File: wallet/src/main/scala/org/bitcoins/wallet/db/WalletDbManagement.scala [hunk 5]
```diff
@@ -1,25 +1,58 @@
...
+      .asInstanceOf[TableQuery[Table[_]]]
+  }
+  private lazy val incomingTxTable: TableQuery[Table[_]] = {
+    IncomingTransactionDAO()(ec, appConfig.asInstanceOf[WalletAppConfig]).table
+      .asInstanceOf[TableQuery[Table[_]]]
+  }
+  private lazy val outgoingTxTable: TableQuery[Table[_]] = {
+    OutgoingTransactionDAO()(ec, appConfig.asInstanceOf[WalletAppConfig]).table
+      .asInstanceOf[TableQuery[Table[_]]]
+  }
```
Comment (benthecarman) [inline wallet/src/main/scala/org/bitcoins/wallet/db/WalletDbManagement.scala]: Don't need to cast `appConfig` in any of these

File: wallet/src/main/scala/org/bitcoins/wallet/db/WalletDbManagement.scala [hunk 6]
```diff
@@ -1,25 +1,58 @@
...
+
+  override lazy val allTables = {
```
Comment (benthecarman) [inline wallet/src/main/scala/org/bitcoins/wallet/db/WalletDbManagement.scala]: nit: give type annotation

File: wallet/src/main/scala/org/bitcoins/wallet/models/AddressDAO.scala [hunk 7]
```diff
@@ -98,6 +112,188 @@ case class AddressDAO()(
...
+       address,
+       scriptWitness,
+       scriptPubKey,
+       addressIndex,
+       ecPublicKey,
+       hashedPubKey,
+       scriptType) <> (fromTuple, toTuple)
+
+
+
```
Comment (benthecarman) [inline wallet/src/main/scala/org/bitcoins/wallet/models/AddressDAO.scala]: nit: remove extra spaces here

File: wallet/src/main/scala/org/bitcoins/wallet/models/AddressDAO.scala [hunk 8]
```diff
@@ -98,6 +112,188 @@ case class AddressDAO()(
...
+    // for some reason adding a type annotation here causes compile error
+    def fk: ForeignKeyQuery[_, AccountDb] =
```
Comment (benthecarman) [inline wallet/src/main/scala/org/bitcoins/wallet/models/AddressDAO.scala]: I guess this comment can be deleted?

File: wallet/src/main/scala/org/bitcoins/wallet/models/SpendingInfoDAO.scala [hunk 9]
```diff
@@ -11,21 +16,29 @@ import org.bitcoins.core.protocol.transaction.{
...
+
+  private lazy val addrTable: slick.lifted.TableQuery[AddressDAO#AddressTable] = {
+    AddressDAO()(ec, appConfig).table
+      .asInstanceOf[TableQuery[AddressDAO#AddressTable]]
+  }
 
+  private lazy val txTable: slick.lifted.TableQuery[IncomingTransactionDAO#IncomingTransactionTable] = {
+    IncomingTransactionDAO().table
+      .asInstanceOf[TableQuery[IncomingTransactionDAO#IncomingTransactionTable]]
+  }
```
Comment (benthecarman) [inline wallet/src/main/scala/org/bitcoins/wallet/models/SpendingInfoDAO.scala]: I don't think we need these casts

**Review by benthecarman — COMMENTED** [review]
TBH I'm not familiar enough with this stuff to give a full review of this, but tests are passing so it works (tm). Gave review on some things that can be fixed up

File: node/src/main/scala/org/bitcoins/node/models/BroadcastAbleTransactionDAO.scala (hunk 3 above)
Comment (Christewart) [reply node/src/main/scala/org/bitcoins/node/models/BroadcastAbleTransactionDAO.scala:55]: I'm not going to do that on this PR as the diff is already big enough, I don't think I got rid of it in `BlockHeaderDAO` it just didn't exist in the first place.

File: node/src/main/scala/org/bitcoins/node/models/PeerDAO.scala (hunk 4 above)
Comment (Christewart) [reply node/src/main/scala/org/bitcoins/node/models/PeerDAO.scala:9]: Yeah, it wasn't implemented. It was essentially a place holder for something that never got implemented.

File: wallet/src/main/scala/org/bitcoins/wallet/models/SpendingInfoDAO.scala (hunk 9 above)
Comment (Christewart) [reply wallet/src/main/scala/org/bitcoins/wallet/models/SpendingInfoDAO.scala]: We do unfortunately, i can't remove them on any tables where we were are referencing another `DAO`

File: node/src/main/scala/org/bitcoins/node/db/NodeDbManagement.scala (hunk 2 above)
Comment (Christewart) [reply node/src/main/scala/org/bitcoins/node/db/NodeDbManagement.scala]: Yes.... this is sooo confusing to me. I don't understand why this won't compile. For reference, if I don't have these casts in all of my `*DbManagement.scala` files I get: 

>[error] /home/suredbits/dev/bitcoin-s/node/src/main/scala/org/bitcoins/node/db/NodeDbManagement.scala:18:35: type mismatch;
[error]  found   : org.bitcoins.db.AppConfig
[error]  required: org.bitcoins.node.config.NodeAppConfig
[error]     BroadcastAbleTransactionDAO()(appConfig, ec).table
[error]                                   ^
[error] one error found

File: db-commons/src/main/scala/org/bitcoins/db/CRUD.scala [hunk 10]
```diff
@@ -1,11 +1,90 @@
...
+  }
+
+  /** The database we are connecting to */
+  lazy val database: Database = {
+    dbConfig.db
+  }
+
+  /** The path where our DB is located */
+  // todo: what happens to this if we
+  // dont use SQLite?
```
Comment (nkohen) [inline db-commons/src/main/scala/org/bitcoins/db/CRUD.scala]: ping for this and below, don't need anymore right?

File: db-commons/src/main/scala/org/bitcoins/db/CRUDAutoInc.scala [hunk 11]
```diff
@@ -1,34 +1,53 @@
...
-    extends CRUD[T, Long] {
+    implicit ec: ExecutionContext,
+    override val appConfig: AppConfig)
+    extends CRUD[T, Long]()(ec, appConfig)
+    with TableAutoIncComponent[T] {
+  import profile.api._
 
   /** The table inside our database we are inserting into */
-  override val table: TableQuery[_ <: TableAutoInc[T]]
+  override val table: profile.api.TableQuery[_ <: TableAutoInc[T]]
```
Comment (nkohen) [inline db-commons/src/main/scala/org/bitcoins/db/CRUDAutoInc.scala:14]: nit: Unnecessary `profile.api` (due to import)

File: node/src/main/scala/org/bitcoins/node/models/BroadcastAbleTransactionDAO.scala [hunk 12]
```diff
@@ -1,21 +1,25 @@
...
+final case class BroadcastAbleTransactionDAO()(
+    implicit override val appConfig: NodeAppConfig,
     val ec: ExecutionContext)
     extends CRUDAutoInc[BroadcastAbleTransaction] {
 
-  val table: TableQuery[BroadcastAbleTransactionTable] =
-    TableQuery[BroadcastAbleTransactionTable]
+  import profile.api._
+
+  val table: profile.api.TableQuery[BroadcastAbleTransactionTable] =
```
Comment (nkohen) [inline node/src/main/scala/org/bitcoins/node/models/BroadcastAbleTransactionDAO.scala]: nit: `override`

File: wallet-test/src/test/scala/org/bitcoins/wallet/WalletAppConfigTest.scala [hunk 13]
```diff
@@ -9,13 +9,16 @@ import org.bitcoins.core.config.RegTest
...
 import org.bitcoins.core.hd.HDPurposes
 import java.nio.file.Files
+
 import ch.qos.logback.classic.Level
 import java.nio.file.Path
+
 import scala.util.Properties
 
-class WalletAppConfigTest extends BitcoinSUnitTest {
+class WalletAppConfigTest extends BitcoinSAsyncTest {
```
Comment (nkohen) [inline wallet-test/src/test/scala/org/bitcoins/wallet/WalletAppConfigTest.scala:21]: why the change? I'm not seeing any `Future`s below?

Comment (nkohen) [reply wallet-test/src/test/scala/org/bitcoins/wallet/WalletAppConfigTest.scala:21]: If it is just to get the `ExecutionContext` probably best to just import that directly

Comment (Christewart) [reply wallet-test/src/test/scala/org/bitcoins/wallet/WalletAppConfigTest.scala:21]: We need to get that execution context from somewhere thought and I don't want to use the scala runtime execution context (scala.concurrent.ExecutionContext.Implicits.global)

File: db-commons/src/main/scala/org/bitcoins/db/DbManagement.scala [hunk 14]
```diff
@@ -1,34 +1,48 @@
...
+    tableQuery.asInstanceOf[TableQuery[Table[_]]]
+  }
+
+  def allTables: List[TableQuery[Table[_]]]
 
   /** Lists all tables in the given database */
-  def listTables(db: Database): Future[Vector[SQLiteTableInfo]] = {
+  def listTables: Future[Vector[SQLiteTableInfo]] = {
     import DbCommonsColumnMappers._
     val query = sql"SELECT * FROM sqlite_master where type='table'"
```
Comment (rorp) [inline db-commons/src/main/scala/org/bitcoins/db/DbManagement.scala]: Is there any way to extract `java.sql.DatabaseMetaData` using `slick`? This query will fail on many other databases.

**Pull request merged**